common/persistence/queue_v2_util.go

```diff
@@ -54,22 +56,87 @@ func GetNextPageTokenForQueueV2(result []QueueV2Message) []byte {
 	return append([]byte{pageTokenPrefixByte}, b...)
 }
 
-func GetMinMessageIDForQueueV2(request *InternalReadMessagesRequest) (int64, error) {
-	// TODO: start from the ack level of the queue partition instead of the first message ID when there is no token.
-	if len(request.NextPageToken) == 0 {
-		return FirstQueueMessageID, nil
+func GetMinMessageIDToReadForQueueV2(
+	queueType QueueV2Type,
+	queueName string,
+	nextPageToken []byte,
+	queue *persistencespb.Queue,
+) (int64, error) {
+	if len(nextPageToken) == 0 {
+		partition, err := GetPartitionForQueueV2(queueType, queueName, queue)
+		if err != nil {
+			return 0, err
+		}
+		return partition.MinMessageId, nil
 	}
 	var token persistencespb.ReadQueueMessagesNextPageToken
 
 	// Skip the first byte. See the comment on pageTokenPrefixByte for more details.
-	err := token.Unmarshal(request.NextPageToken[1:])
+	err := token.Unmarshal(nextPageToken[1:])
 	if err != nil {
 		return 0, fmt.Errorf(
 			"%w: %q: %v",
 			ErrInvalidReadQueueMessagesNextPageToken,
-			request.NextPageToken,
+			nextPageToken,
 			err,
 		)
 	}
 	return token.LastReadMessageId + 1, nil
 }
+
+func GetPartitionForQueueV2(
+	queueType QueueV2Type,
+	queueName string,
+	queue *persistencespb.Queue,
+) (*persistencespb.QueuePartition, error) {
+	// Currently, we only have one partition for each queue. However, that might change in the future. If a queue is
+	// created with more than 1 partition by a server on a future release, and then that server is downgraded, we
+	// will need to handle this case. Since all DLQ tasks are retried infinitely, we just return an error.
+	numPartitions := len(queue.Partitions)
+	if numPartitions != 1 {
+		return nil, serviceerror.NewInternal(
+			fmt.Sprintf(
+				"queue without single partition detected. queue with type %v and queueName %v has %d partitions, "+
+					"but this implementation only supports queues with 1 partition. Did you downgrade your Temporal server?",
+				queueType,
+				queueName,
+				numPartitions,
+			),
+		)
+	}
+	partition := queue.Partitions[0]
+	return partition, nil
+}
+
+type DeleteRequest struct {
+	// LastIDToDeleteInclusive represents the maximum message ID that the user wants to delete, inclusive.
+	LastIDToDeleteInclusive int64
+	// ExistingMessageRange represents an inclusive range of the minimum message ID and the maximum message ID in the queue.
+	ExistingMessageRange InclusiveMessageRange
+}
+
+type InclusiveMessageRange struct {
+	MinMessageID int64
+	MaxMessageID int64
+}
+
+type DeleteRange struct {
+	InclusiveMessageRange
+	NewMinMessageID int64
+}
+
+// GetDeleteRange returns the range of messages to delete, and a boolean indicating whether any messages should be deleted.
+func GetDeleteRange(request DeleteRequest) (DeleteRange, bool) {
```

Comment:
This implementation looks **much** better. I was thinking of some test cases for this, and I realized that the boolean we return doesn't exactly mean there are messages to delete. We might just update min message ID without deleting any messages. For example:

```go
DeleteRequest{
  LastIDToDeleteInclusive: 1,
  ExistingMessageRange: InclusiveMessageRange{
    MinMessageID: 1,
    MaxMessageID: 1,
  },
}
```

should return:

```go
DeleteRange{
  InclusiveMessageRange: InclusiveMessageRange{
    MinMessageID: 1,
    MaxMessageID: 0,
  },
  NewMinMessageID: 2,
}, true
```

I think we should clarify in the comment that it returns true if the min message ID should be updated, something like this:

```go
// GetDeleteRange returns the range of messages to delete, and a boolean indicating whether there is any update to be
// made: meaning either we should delete messages, update the min message ID, or both.
```

Another thing. It's kind of hard to write integration tests for this function, but it's quite easy to unit test it. I asked ChatGPT to come up with a bunch, and then I edited them here: https://gist.github.com/MichaelSnowden/77c1e5fee5af92ec55f202d61c435ce3

I think we should add these to a unit test just for this file.

Reply:
That is a good idea! Added the tests.
Also updated the comment.